use-cases/vehicle-routing-time-windows/src/main/java/org/acme/vehiclerouting/domain/geo/HaversineDistanceCalculator.java

```diff
@@ -0,0 +1,45 @@
+package org.acme.vehiclerouting.domain.geo;
+
+import org.acme.vehiclerouting.domain.Location;
+
+public class HaversineDistanceCalculator implements DistanceCalculator {
+
+    private static final int EARTH_RADIUS_IN_KM = 6371;
+    private static final int TWICE_EARTH_RADIUS_IN_KM = 2 * EARTH_RADIUS_IN_KM;
+
+    @Override
+    public long calculateDistance(Location from, Location to) {
+        if (from.equals(to)) {
+            return 0L;
+        }
+
+        CartesianCoordinate fromCartesian = locationToCartesian(from);
+        CartesianCoordinate toCartesian = locationToCartesian(to);
```

Comment:
If it overwrite the bulk method, it can do the locationToCartesian once for every location, instead of doing it n times for n locations. With 10k locations, I believe this could make a noticable difference. We should measure don't guess.

Reply:
The benchmark is on the way, but I would not block this PR with that. I am happy to add caching after we have the data.

Reply:
It wouldn't need caching I am thinking. The CartesianCoordinate array would be a local field. To discuss